internal/state/documents.go

```diff
@@ -0,0 +1,166 @@
+package state
+
+import (
+	"log"
+	"time"
+
+	"github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/terraform-ls/internal/document"
+	"github.com/hashicorp/terraform-ls/internal/source"
+)
+
+type DocumentStore struct {
+	db        *memdb.MemDB
+	tableName string
+	logger    *log.Logger
+
+	// TimeProvider provides current time (for mocking time.Now in tests)
+	TimeProvider func() time.Time
+}
+
+func (s *DocumentStore) OpenDocument(dh document.Handle, langId string, version int, text []byte) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
```

Comment:
We basically just rely on `Abort()` being no-op after successful `Commit()`:
https://github.com/hashicorp/go-memdb/blob/542a5804dedaaddf1a0cad8b6a7b0b3fcf438776/txn.go#L112-L115

This in combination with `defer` means we don't have to duplicate `Abort` in every `return` branch - i.e. we don't end up inadvertently introducing a bug.

We have to _either_ `Commit` or `Abort` any write transaction, otherwise the (internal) write mutex lock remains locked, effectively causing a deadlock and preventing any other future write transactions to hang.

It is true that this is not clear from the relevant upstream comment/docs though - I raised a PR to address that: https://github.com/hashicorp/go-memdb/pull/120